### PR TITLE
New Installation Script

### PR DESCRIPTION
**Pull request by n3oxmind**
Hi nana,
based on your directory structure and TODO list: I have create this script with the following features:
  - install any theme variant e.g. 
           ./materia install --color dark --size compact
           ./materia install --size compact ..etc
  - change asset color and render asset color
           ./materia assets --fg "#xxxxxx" --accent "#xxxxxx" ...
           ./materia render-assets 
I also have a template theme that allows you to modify almost everything of your theme. but its not organized, once I have time i'll update this script.
let me know about the changes 
BTW, this is my first contribution on github ((I hope I'm doing it the right way)

### CODE REVIEW

**Comment by actionless** [comment]
it's breaking functionality of pseudo-hidpi gtk2 generation:

https://github.com/nana-4/materia-theme/blob/master/src/gtk-2.0/render-asset.sh#L12-L17

also `./materia install` is seems to be missing the flow of `./change_color.sh` -- to install a theme variant with specified colors into a custom location

**Comment by n3oxmind** [comment]
I can fix pseudo-hidpi gtk2 and soon I'll include the other installation options like the one you specified. e.g. `./materia install --size compact --fg #xxxxxx --bg #xxxxxx --accent #xxxxxx --opacity 0.8`. (this way you can make it even more customizable ). 
this is just the start.. please, let me know if there is another options or a feature i'm missing.

**Comment by nana-4** [comment]
Oops, I overlooked this. Sorry! (When I tried to see this page before, it was 404 not found...)

BTW, I really appreciate your great work!!!

For details, I'll review later, but roughly speaking;

1. I still need `test.sh` or similar functionality for the theme development. `test.sh` makes it possible to check the changes quickly by using _symlink_ (`ln -s`) instead of _copy_ (`cp`).

2. `parse-sass.sh` and `render-assets.sh` scripts are also needed to contain the compiled css/png in the repository. I want to keep them in the repo to reduce the user's burden (for installation time and less dependency).

3. `assets` and `render-assets` options seem to recolor only `*.png` assets, but not `*.css` and `gtk-2.0/gtkrc*`. Also, such function is already realized with `change_color.sh`. So I'm not sure those options are really necessary. (Or, it needs to support `*.css` and `gtkrc` (and also ideally other `*.svg`) to enhance the function.)

**Comment by actionless** [comment]
also, to rephrase/append_to my previous comment:
i would recommend you to try using `./change_color.sh` with few different color values to see the expected result of it.

you can also use `oomox` app to get more colorschemes for testing:
https://github.com/actionless/oomox/tree/master/colors

so, i think that's quite important for your refactored script version to still produce the same resulting gtk theme as the one generated with `./change_color.sh` script when running with the same colors

File: materia.sh [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+    printf "  %-20s%s\n" "assets" "Changing assets color"
+    printf "  %-20s%s\n" "render-assets" "Rendering assets color" 
+    printf "\n%s\n" "OPTIONS:"
+    printf "  %-20s%s\n" "-c, --color" "Specify theme color variant (Default to ALL variants)"
+    printf "  %-20s%s\n" "-s, --size" "Specify theme size variant (Default to ALL sizes)"
+    printf "  %-20s%s\n" "-n, --name" "Specify theme name (Default to Materia-color-size)"
+    printf "  %-20s%s\n" "-d, --dst" "Specify theme destination folder (Default to ~/.themes)"
+    printf "  %-20s%s\n" "-h, --help" "Show this help"
+    #printf "  %-20s%s\n" "asset" "Change assets color(Only takes ASSETS OPTIONS as an arg)"
+    #printf "  %-20s%s\n" "render-assets" "Rendering assets (Default to ouput PNGs with 90&192 dpi)"
```
Comment (nana-4) [inline materia.sh]: Typo: `90&192` -> `96&192`

File: materia.sh [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+    install -d "${gs_dst_dir}"
+    cp -ur ${gs_src_dir}/*.svg ${gs_dst_dir}
+    cp -urL ${gs_src_dir}/{extensions,pad-osd.css} ${gs_dst_dir}
+    if [ "${1}" != "-dark" ]; then
+        cp -urL ${gs_src_dir}/assets ${gs_dst_dir}
+    else
+        cp -urL ${gs_src_dir}/assets-dark ${gs_dst_dir}/assets
+    fi
+    # Compile .scss files to .css files 
+    sass "${gs_src_dir}/gnome-shell${1}${2}.scss" "${gs_dst_dir}/gnome-shell.css"
```
[the remark this answers is not in the document]
Comment (actionless) [reply materia.sh]: good point, it's much faster and have less dependencies

**Comment by nana-4** [comment]
In addition, I personally feel it is better to divide the _recolor_ functions (around `asset` & `render-assets`) into the another script (e.g. named `recolor`). The reasons are current `materia.sh` code is a bit too long, and the _recolor_ feature would work as _optional_.

**Comment by nana-4** [comment]
Also I'd like to talk my thoughts about changing the color/opacity. Maybe they are quite hard to realize (especially No.3) but please forgive me to say this. :)

1. `./install.sh --primary #0000FF --accent #FF0000 --bg #EEEEEE` should be better to change _only_ the specified colors, shouldn't change the unspecified colors.

2. **Background colors** (including `primary` and `accent`) with _transparency_ has some technical problems. (e.g. GTK2 (`gtkrc`) doesn't support transparency colors except for assets.) Therefore, I think opacity should not be controllable for the background colors.

3. For **text and similar colors**, the best way is that `./install.sh --fg #FF0000` will change `color-fg` and similar colors based on the [Material Design color theory](https://material.io/guidelines/style/color.html#color-usability) as follows:

  | color vars | new color | new opacity |
  |-|-|-|
  | `color-fg` | `#FF0000` | `1` |
  | `color-fg-secondary` | `#FF0000` | `0.7` |
  | `color-fg-hint` | `#FF0000` | `0.5` |
  | `color-fg-disabled` | `#FF0000` | `0.5` |
  | `color-fg-secondary-disabled` | `#FF0000` | `0.3` |
  | `color-track` | `#FF0000` | `0.3` |
  | `color-divider` | `#FF0000` | `0.12` |

  _(These opacities are same as the `-dark` values.)_

File: materia.sh [hunk 3]
```diff
@@ -0,0 +1,625 @@
+#!/bin/env bash
+set -e
+if [ $# -eq 0 ]; then
+    echo "$0: missing argument"
+    echo "Try '$0 --help' for more information."
+    exit 1
+fi
+REPO_DIR=$(cd $(dirname $0) && 'pwd')
+SRC_DIR=${REPO_DIR}/src
+THEME_DIR=$HOME/.themes
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: thanks for the review will correct that.

File: materia.sh [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+    printf "\n%s\n" "ASSETS OPTIONS:"
+    printf "  %-25s%s\n" "--fg" "Change assets color-fg"
+    printf "  %-25s%s\n" "--fg-dark" "Change assets color-fg-dark"
+    printf "  %-25s%s\n" "--bg" "Change assets color-bg"
+    printf "  %-25s%s\n" "--bg-dark" "Change assets color-bg-dark"
+    printf "  %-25s%s\n" "--bg-lighter" "Change assets color-bg-lighter"
+    printf "  %-25s%s\n" "--bg-lighter-dark" "Change assets color-bg-lighter-dark"
+    printf "  %-25s%s\n" "--base" "Change assets color-base"
+    printf "  %-25s%s\n" "--base-dark" "Change assets color-base-dark"
+    printf "  %-25s%s\n" "--accent" "Change assets color-accent"
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: I will check that out and fix it.

File: materia.sh [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+    printf "\n" 
+    printf "%s\n" "INSTALLATION EXAMPLES:"
+    printf "%s\n" "Install all theme variants with default name"
+    printf "\t%s\t%s\n" "./materia install"
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: thanks for that, I'll update the 'usage'.

File: materia.sh [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+    && sleep $f && s=`echo ${s} + ${f} + ${f} | bc` && echo -ne "\r\t\t\t\t\t[########   ] [70%]." \
+    && sleep $f && s=`echo ${s} + ${f} + ${f} | bc` && echo -ne "\r\t\t\t\t\t[#########  ] [80%]." \
+    && sleep $f && s=`echo ${s} + ${f} + ${f} | bc` && echo -ne "\r\t\t\t\t\t[########## ] [90%]." ; 
+       sleep $f && s=`echo ${s} + ${f} + ${f} | bc` && echo -ne "\r\t\t\t\t\t[###########] [100%].";
+    done
+    echo -ne "\n"
+}
+# Get gnome-shell version
+if [[ $(which gnome-shell 2> /dev/null) ]]; then
+    gnome_shell_ver=$(gnome-shell --version | cut -d ' ' -f 3 | cut -d . -f 1,2)
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: nice one, I'll fix it.

File: materia.sh [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+        change_asset_color "color-accent" "${color_accent}"
+        change_asset_color "color-accent-disabled" "${color_accent}"
+    fi
+    print_assets_summary
+    echo 
+    echo "Assets color successfully changed, run '${0} render-assets' to generate new assets."
+}
+print_assets_summary () {
+    printf "%-25s%-10s%-10s%-8s\n" "ASSETS COLOR" "OLD COLOR" "NEW COLOR" "OPACITY"
+    printf "%-25s%-10s%-10s%-8s\n" "color-fg" "${old_color_fg}" "${color_fg}" "0.78"
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: all typo will be fixed in the next update.

File: materia.sh [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+                exit 1
+                ;; 
+        esac
+    done
+    change_assets_color
+    echo
+elif [ "${1}" == "render-assets" ]; then
+    shift
+    while [ $# -ne 0 ] && [ "$1" != "--" ]; do
+        case "$1" in
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: I was testing gtk2 and gtk3 separately, the default option will be installing both.
thank you for that.

**Comment by n3oxmind** [comment]
I'll check this function again and see what its missing other than --primary color. about opacity I haven't included yet in the script.

File: materia.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: sassc is much faster than sass, thanks for this.

File: materia.sh [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+}
+# Installing GTK+ 3 Theme 
+install_gtk3_theme () {
+    install -d "${3}/gtk-common"
+    cp -ur ${SRC_DIR}/gtk-3.0/gtk-common/assets ${3}/gtk-common
+    for ver in "${gtk_ver[@]}"; do 
+        if [ "${ver}"  == "3.18" ]; then
+            install -d "${3}/gtk-3.0"
+            cp -ur ${SRC_DIR}/gtk-3.0/${ver}/assets ${3}/gtk-3.0
+            sass "${SRC_DIR}/gtk-3.0/${ver}/gtk${1}.scss" "${3}/gtk-3.0/gtk.css" &>/dev/null
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: I'll add that

File: materia.sh [hunk 10]
```diff
@@ -0,0 +1,625 @@
...
+        && $OPTIPNG -o7 --quiet ${assets_dir}/${asset_name}@2.png
+    fi
+}
+# Rendering assets
+render_assets () {
+    if [ "$1" == "gtk3" ]; then
+        if [ ! -d "$SRC_DIR/gtk-3.0/gtk-common/assets" ]; then
+            mkdir -p "${SRC_DIR}/gtk-3.0/gtk-common/assets"
+        elif [ ! -d "$SRC_DIR/gtk-3.0/gtk-common/assets_original" ]; then
+            mv "$SRC_DIR/gtk-3.0/gtk-common/assets" "$SRC_DIR/gtk-3.0/gtk-common/assets_original"
```
[the remark this answers is not in the document]
Comment (n3oxmind) [reply materia.sh]: nice once, I'll fix it.

**Comment by nana-4** [comment]
Hi, @n3oxmind 
First of all, thank you for your hard work! I'm happy to see your elaborate install script.

However I have some questions:

1. Currently **228** files are changed. `.gitignore`, `HACKING.md`, etc. are also deleted. Is it intentional? If not, could you revert the files that is no-need to change?

2. This is a repetition/paraphrase of my previous comment; When installing **vanilla** Materia, IMHO it is better to _just_ copy pre-compiled `.css` files without using `sassc` command **in order to reduce installation time and dependency**. What do you think about this? (FYI, [Adwaita](https://github.com/GNOME/gtk/tree/master/gtk/theme/Adwaita), [Arc](https://github.com/horst3180/arc-theme/tree/master/common/gtk-3.0/3.20), etc. also have pre-compiled `.css` files in their own repositories.)

**Comment by n3oxmind** [comment]
1- I don't mind using pre-compiled `.css' files. but by using `sassc' you don't need all of the `gtk-color-size.css`,`gtk-color-size.scss` , `gnome-shell-color-size.css` and 'gnome-shell-color-size.scss' the script will generate the proper file whenever needed. 
2- bring back all of the `.css` & `.scss` file will not break the script at all. I'll make the script do the following (if you want):
         1- For normal installation (without modifying the original theme) use copy.
         2- For custom installation (modifying the original theme) use `sassc`.
3- About .gitignore and HACKING.md I'll put them back.

**Comment by nana-4** [comment]
> but by using sassc you don't need all of the gtk-color-size.css,gtk-color-size.scss , gnome-shell-color-size.css and gnome-shell-color-size.scss the script will generate the proper files whenever needed.

yeah, but I need all of the compiled `gtk${color}${size}.css` and `gnome-shell${color}${size}.css` for **easier testing** (theme development) by using `ln -s`. (Please check [`test.sh`](https://github.com/nana-4/materia-theme/blob/master/test.sh).)

> bring back all of the .css & .scss files will not break the script at all. I'll make the script do the following (if you want):

Sounds/looks great. Thanks!

**Comment by nana-4** [comment]
Hello @n3oxmind,

I really hate to say this because it might mess up your kind hard work so far, but...
After reconsideration, I thought there are several (bigger) issues to be solved.

1. The `materia.sh` code is too complicated **for me to maintain**.

    Actually, I'm not as familiar with shell script as you. Also for people who fork this theme, I think the installation script should be as simple as possible. (To be honest, I was trying to make a simplified version of `install.sh` before this PR. Please see the link below.)

    The best solution I think is to separate the recolor feature from the installation script (like `install.sh` and `change_color.sh`). This would also be helpful if we migrate to `make` or `meson` in the future.

2. The following files are still necessary for quick testing using `./test.sh`.
  - `parse-sass.sh` (This is also necessary to contain the pre-compiled CSSs in the repo.)
  - `render-assets.sh` (Although these _render-assets_ scripts can/should be simplified...)
    - `src/gtk-2.0/render-asset.sh`
    - `src/gtk-2.0/render-assets-dark.sh`
    - `src/gtk-2.0/render-assets.sh`
    - `src/gtk-3.0/gtk-common/render-asset.sh`
    - `src/gtk-3.0/gtk-common/render-assets.sh`
  - `src/index*.theme` (Using `sed` looks interesting, but it breaks my workflow using symlinks.)
3. Current `materia.sh` recolor feature needs to specify colors **separately** for _assets_ (svg/png) and _others_ (scss/css/gtkrc).

    IMO it would be better for users to be able to specify colors **at the same time** like `./change_color.sh`.

    Also, `./materia.sh assets` can preview the color strings before generating the assets, but I wonder a bit if the feature is really necessary... Because users can use the (GUI) [oomox](https://github.com/actionless/oomox) instead of the (CUI) script.

4. There are many conflicting files. (This is not big issue than the above, though.)


---

Based on these issues, I propose the following plan.

1. Close this PR for now (because big changes are needed to solve the above issues).
2. Merge the [simplified version of `install.sh` I made](https://github.com/nana-4/materia-theme/blob/improve-install-sh/install.sh) into `master` branch.
  - This is based on `test.sh` and your work. :)
  - This is still incomplete and needs some bug fixes. (Please see the [commit massage](https://github.com/nana-4/materia-theme/commit/960249da9859908a09f84fa0e499857a7dc6f74c) for details.)
  - In the meantime, I may improve the directory structure (#170) and update the script accordingly.
3. If you still could help us, please improve/fix the (updated) scripts!!
  - For installation scripts, please do not include the recolor functions (if you could agree with my opinion).
  - For the recolor feature, you can simply improve `change_color.sh` or create a new script.
  - When sending PRs, I suggest that you create a branch (not `master`) from the next time, btw. :)

---

What do you think about these issues/plan?
Please do not hesitate to tell me your opinions (and/or objections).
Of course, opinions from other than @n3oxmind are also very welcomed.

**Pull request closed without merging**